Library/Homebrew/utils.rb

```diff
@@ -770,3 +770,24 @@ def number_readable(number)
   (numstr.size - 3).step(1, -3) { |i| numstr.insert(i, ",") }
   numstr
 end
+
+def truncate_text_to_approximate_size(s, max_bytes, options = {})
+  front_weight = options[:front_weight] || 0.5
+  glue = options[:glue] || "\n[...snip...]\n"
+  return s if s.bytesize <= max_bytes
+
+  if front_weight < 0 || front_weight > 1
+    raise "opts[:front_weight] must be between 0.0 and 1.0"
+  end
+  scale = 0.99 # just a tad small to allow for the glue
+  while true
```

Comment:
> I think in this case as we're picking a relatively arbitrary/safe 1MB anyway I'd just restrict this to assuming the worst-case scenario of every UTF-8 character being maximum length and just truncating more than necessary if needed.

That gets us down to about 128 KB max per log file (1 MB per file to prevent unicorns and allow for multiple log files / 8 bytes max encoding per character), which is feeling kind of small for me. And yeah, I know I was one of the folks advocating for truncation in the first place.

> Another option is reencoding this in a fixed-length encoding.

Given the existence of surrogate pairs, you'd have to either use UTF-32 and accept a 4x size error in the vast majority of our use cases (making the effective limit 256 KB per file instead of 1 MB) or accept possible truncation-related encoding errors (and given our nerdy and Asian-leaning user base, that's not inconceivable). Still better than the worst-case UTF-8 scenario, though.

I guess a 4x error storage size error should be negligible in any healthy system, but I feel like that's a lot to pay for getting rid of one level of looping and four or so lines of code.

Wish I had a) stats, and b) a full record of our bottle builds for the last few months so we could see what log file sizes in the wild are like.